### PR TITLE
Use Java 7 language/library features for clarity and simplicity

### PR DESCRIPTION
**Pull request by liblit**
These changes put various Java 7 language and library features to use for improved code clarity and simplicity. All were found and applied using various IntelliJ IDEA inspections, followed by manual examination (and in a few cases. reversal) of the changes.

### CODE REVIEW

File: com.ibm.wala.cast.java/src/com/ibm/wala/cast/java/examples/ast/SynchronizedBlockDuplicator.java [hunk 1]
```diff
@@ -76,7 +77,7 @@ public boolean equals(Object o) {
       return (o instanceof UnwindKey)
           && ((UnwindKey) o).testDirection == testDirection
           && ((UnwindKey) o).syncNode == syncNode
-          && (rest == null ? ((UnwindKey) o).rest == null : rest.equals(((UnwindKey) o).rest));
+          && Objects.equals(rest, ((UnwindKey) o).rest);
```
Comment (msridhar) [inline com.ibm.wala.cast.java/src/com/ibm/wala/cast/java/examples/ast/SynchronizedBlockDuplicator.java]: 👍 I didn't know about this!

File: com.ibm.wala.cast.java.test/src/com/ibm/wala/cast/java/test/JavaIRTests.java [hunk 2]
```diff
@@ -606,13 +606,7 @@ public void testWhileTest1() throws IllegalArgumentException, CancelException, I
...
     try {
       runTest(singleTestSrc(), rtJar, simpleTestEntryPoint(), emptyList, true, null);
-    } catch (IllegalArgumentException e) {
-      // TODO Auto-generated catch block
-      e.printStackTrace();
-    } catch (CancelException e) {
-      // TODO Auto-generated catch block
-      e.printStackTrace();
-    } catch (IOException e) {
+    } catch (IllegalArgumentException | IOException | CancelException e) {
```
Comment (msridhar) [inline com.ibm.wala.cast.java.test/src/com/ibm/wala/cast/java/test/JavaIRTests.java]: 👍

File: com.ibm.wala.cast/source/java/com/ibm/wala/cast/ir/ssa/SSAConversion.java [hunk 3]
```diff
@@ -740,12 +740,8 @@ protected boolean skip(int i) {
       ssa.perform();
       if (DEBUG_UNDO) System.err.println(("<<< done " + ir.getMethod()));
       return ssa.getComputedLocalMap();
-    } catch (RuntimeException e) {
-      //      System.err.println(("exception " + e + " while converting:"));
-      //      System.err.println(ir);
-      throw e;
-    } catch (Error e) {
-      //      System.err.println(("error " + e + " while converting:"));
+    } catch (RuntimeException | Error e) {
```
Comment (msridhar) [inline com.ibm.wala.cast/source/java/com/ibm/wala/cast/ir/ssa/SSAConversion.java]: Can't we just delete this try-catch?  It's just re-throwing

File: com.ibm.wala.shrike/src/com/ibm/wala/shrikeCT/BootstrapMethodsReader.java [hunk 4]
```diff
@@ -146,17 +146,12 @@ public Object callArgument(ClassLoader cl, int i) {
...
-              } catch (InvalidClassFileException e) {
-                assert false : e;
-              } catch (ClassNotFoundException e) {
-                assert false : e;
-              } catch (NoSuchMethodException e) {
-                assert false : e;
-              } catch (SecurityException e) {
-                assert false : e;
-              } catch (IllegalAccessException e) {
+              } catch (IllegalArgumentException
```
Comment (msridhar) [inline com.ibm.wala.shrike/src/com/ibm/wala/shrikeCT/BootstrapMethodsReader.java]: I don't like the use of `assert` here.  Can we re-throw instead, wrapping in a `RuntimeException`?

File: com.ibm.wala.cast.java/src/com/ibm/wala/cast/java/examples/ast/SynchronizedBlockDuplicator.java (hunk 1 above)
Comment (liblit) [reply com.ibm.wala.cast.java/src/com/ibm/wala/cast/java/examples/ast/SynchronizedBlockDuplicator.java]: Yeah, `java.util.Objects.equals` is a nice little helper for a very common code pattern.

File: com.ibm.wala.cast/source/java/com/ibm/wala/cast/ir/ssa/SSAConversion.java (hunk 3 above)
Comment (liblit) [reply com.ibm.wala.cast/source/java/com/ibm/wala/cast/ir/ssa/SSAConversion.java]: Presumably this clause used to be considered useful when it printed an error message, which is now commented out. I will remove the entire `try`–`catch` per your suggestion.

File: com.ibm.wala.shrike/src/com/ibm/wala/shrikeCT/BootstrapMethodsReader.java (hunk 4 above)
Comment (liblit) [reply com.ibm.wala.shrike/src/com/ibm/wala/shrikeCT/BootstrapMethodsReader.java]: Done.

**Pull request merged**